### PR TITLE
EVA-2366 - Detect merge type and implement basic horizontal merge

### PR DESCRIPTION
**Pull request by apriltuesday**

### CODE REVIEW

File: eva_vcf_merge/detect.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    SINGLE_SET = 'single set'
+    UNSORTED_SINGLE_SET = 'unsorted single set'
+    UNIQUE_SETS = 'unique sets'
+    OVERLAPPING_SETS = 'overlapping sets'
+
+
+class MergeType(Enum):
+    HORIZONTAL = 'horizontal'
+    VERTICAL = 'vertical'
+
```
Comment (sundarvenkata-EBI) [inline eva_vcf_merge/detect.py:33]: More of a nice to have: a brief description of each of the SampleSetType and also an illustration (or a URL to an existing one) of horizontal and vertical merging.

File: eva_vcf_merge/detect.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+def get_samples_from_vcf(vcf_file):
+    """Get the list of samples present in a single VCF file."""
+    with pysam.VariantFile(vcf_file, 'r') as vcf_in:
+        samples = list(vcf_in.header.samples)
+    return samples
+
+
+def compare_sample_sets(list_of_sample_names):
```
Comment (sundarvenkata-EBI) [inline eva_vcf_merge/detect.py:62]: A description of the shape of the data would be useful here. My understanding is that this is a list of list of sample names in each VCF.

File: eva_vcf_merge/detect.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    """
+    file_to_sample_names = {}
+    # retrieve all the sample_names from the VCF files
+    for file_path in vcf_files:
+        file_to_sample_names[file_path] = get_samples_from_vcf(file_path)
+    # Check that all the samples are the same and in the same order to enable horizontal merging
+    sample_info = compare_sample_sets(file_to_sample_names.values())
+    if sample_info == SampleSetType.UNIQUE_SETS:
+        return MergeType.HORIZONTAL
+    elif sample_info == SampleSetType.SINGLE_SET:
```
Comment (sundarvenkata-EBI) [inline eva_vcf_merge/detect.py:50]: Did bcftools merge have issues VCFs with UNSORTED_SINGLE_SET i.e., swapped sample order?

File: tests/test_merge.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def test_horizontal_merge_multiple_groups(vcf_merger, unique_samples_vcfs, unique_samples_vcfs_2):
+    vcfs = {'1': unique_samples_vcfs, '2': unique_samples_vcfs_2}
+    filenames = vcf_merger.horizontal_merge(vcfs)
+    assert sorted(filenames) == [
+        os.path.join(vcf_merger.output_dir, '1_merged.vcf.gz'),
+        os.path.join(vcf_merger.output_dir, '2_merged.vcf.gz')
+    ]
+    assert_all_files_present(filenames)
```
Comment (sundarvenkata-EBI) [inline tests/test_merge.py:22]: This is a broader question for further discussion: should we have tests to ensure that the final merged file has the INFO fields and other data points (genotypes etc.,) correctly populated? A point that was brought up in a recent stand-up was that "bcftools merge" (or concat for that matter) could sometimes compromise the integrity of data in the source files. Also, Thomas brought up a point about dealing with multi-allelics. How do we address such issues?

**Review by sundarvenkata-EBI — APPROVED** [review]

File: eva_vcf_merge/detect.py (hunk 3 above)
Comment (apriltuesday) [reply eva_vcf_merge/detect.py:50]: Do you mean bcftools concat (i.e. "vertical merge")?  The [docs for concat](http://samtools.github.io/bcftools/bcftools.html#concat) call for "the same sample columns appearing in the same order" but to be honest I haven't actually tried it.

File: eva_vcf_merge/detect.py (hunk 1 above)
Comment (apriltuesday) [reply eva_vcf_merge/detect.py:33]: I added some description in the code and some (very simple) illustration to the [readme](https://github.com/apriltuesday/eva-vcf-merge/blob/EVA-2366/README.md), can you let me know if it seems useful?

If we want to make this really useable by other people, we will probably need more examples and diagrams that aren't just colourful boxes, and some detail on the issues you describe in your [other comment](https://github.com/EBIvariation/eva-vcf-merge/pull/2/files#r685468202)... but I guess we have to start somewhere.

Comment (sundarvenkata-EBI) [reply eva_vcf_merge/detect.py:33]: Wow! I was just looking for a poor-man's stick diagram. This is way better!! Thanks!

File: eva_vcf_merge/detect.py (hunk 3 above)
Comment (sundarvenkata-EBI) [reply eva_vcf_merge/detect.py:50]: No worries. Yes, I did mean concat. Even if it works, it is good to not try undocumented things. So I agree with the decision here to bail on such a scenario.

**Comment by apriltuesday** [comment]
This PR stopped picking up my new commits due to some fork shenanigans, please take a look at #3 instead.

**Pull request closed without merging**